src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +267,355 @@ create_distributed_table(PG_FUNCTION_ARGS)
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	RangeVar *rangeVar = makeRangeVar(schemaName, tableName, -1);
+
+	if (!IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+	{
+		EnsureTableNotDistributed(relationId);
+
+		/*
+		 * Before taking locks, convert the table into a Citus local table and commit
+		 * to allow shard split to see the shard.
+		 */
+		char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+
+		StringInfo command = makeStringInfo();
+		appendStringInfo(command,
+						 "SELECT pg_catalog.citus_add_local_table_to_metadata(%s)",
+						 quote_literal_cstr(qualifiedRelationName));
+
+		ExecuteRebalancerCommandInSeparateTransaction(command->data);
+	}
+
+	/*
+	 * Lock target relation with a shard update exclusive lock to
+	 * block DDL, but not writes.
+	 *
+	 * If there was a concurrent drop/rename, error out by setting missingOK = false.
+	 */
+	bool missingOK = false;
+	relationId = RangeVarGetRelid(rangeVar, ShareUpdateExclusiveLock, missingOK);
+
+	if (PartitionedTable(relationId))
+	{
+		/* also lock partitions */
+		LockPartitionRelations(relationId, ShareUpdateExclusiveLock);
+	}
+
+	List *shardList = LoadShardIntervalList(relationId);
+
+	/*
+	 * It's technically possible for the table to have been concurrently
+	 * distributed just after citus_add_local_table_to_metadata and just
+	 * before acquiring the lock, so double check.
+	 */
+	if (list_length(shardList) != 1 ||
+		!IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("table was concurrently modified")));
+	}
+
+	/*
+	 * The table currently has one shard, we will split that shard to match the
+	 * target distribution.
+	 */
+	ShardInterval *shardToSplit = (ShardInterval *) linitial(shardList);
+
+	PropagatePrerequisiteObjectsForDistributedTable(relationId);
+
+	bool viaDeprecatedAPI = false;
+	char replicationModel = DecideReplicationModel(distributionMethod,
+												   colocateWithTableName,
+												   viaDeprecatedAPI);
+
+	Var *distributionColumn = BuildDistributionKeyFromColumnName(relationId,
+																 distributionColumnName,
+																 NoLock);
+
+	Oid distributionColumnType = distributionColumn->vartype;
+	Oid distributionColumnCollation = distributionColumn->varcollid;
+
+	/*
+	 * At this stage, we only want to check for an existing co-location group.
+	 * We cannot create a new co-location group until after replication slot
+	 * creation in NonBlockingShardSplit.
+	 */
+	uint32 colocationId = FindColocateWithColocationId(relationId,
+													   replicationModel,
+													   distributionColumnType,
+													   distributionColumnCollation,
+													   shardCount,
+													   shardCountIsStrict,
+													   colocateWithTableName);
+
+	EnsureRelationCanBeDistributed(relationId, distributionColumn, distributionMethod,
```

Comment:
it would be nice if we could do this check before citus_add_local_table_to_metadata to not do anything in the error case, though we should avoid holding any locks on the relation ID